### PR TITLE
[v7r2] Add Extensions module and make integration tests easier to run

### PR DESCRIPTION
**Pull request by chrisburr**
The commits before 5b19049 are from other PRs which are needed to make the CI pass in this one. The correct diff can be found at: https://github.com/DIRACGrid/DIRAC/compare/5b190494da7f50332c699aa0baeaf3d5aa401f39...chrisburr:new-int-tests?expand=1

* d0769718b5e10e970b9b69034bbac39fb44590d1: Stop using DIRACOSVER=https for HTTP integration tests
* 69877a91c4f5962000b707fbef78f89aa9a025c1: Create `DIRAC.Core.Utilities.Extensions` module. This moves some of the helpers from `DIRAC.Core.Utilities.DIRACScript` and they're also used by the tests to find the agents/databases. #5000 includes more invasive changes to use the new functions in the ComponentInstaller.
* 954686ca5f4c6fd09fb19b474381dcba26c58c70: New python based integration test script. This makes it easier to run the integration tests by replacing the main shell script with a python script. It also makes it much easier to run the CI for extensions ([example for LHCbDIRAC](https://gitlab.cern.ch/cburr/LHCbDIRAC/-/blob/b5e81f3237232ba4f13114594e65662678a73913/.gitlab-ci.yml#L182-221)). I think the `./integration_tests.py --help` sums it up:

```
Usage: integration_tests.py [OPTIONS] COMMAND [ARGS]...

  Run the DIRAC integration tests.

  A local DIRAC setup can be created and tested by running:

    ./integration_tests.py create

  This is equivalent to running:

    ./integration_tests.py prepare-environment
    ./integration_tests.py install-server
    ./integration_tests.py install-client
    ./integration_tests.py test-server
    ./integration_tests.py test-client

  The test setup can be shutdown using:

    ./integration_tests.py destroy

  See below for additional subcommands which are useful during local
  development.

  ## Features

  The currently known features and their default values are:

    HOST_OS: 'cc7'
    MYSQL_VER: '8.0'
    ES_VER: '7.9.1'
    DIRACOSVER: 'master'
    DIRACOS_TARBALL_PATH: None
    TEST_HTTPS: 'No'
    DIRAC_USE_NEWTHREADPOOL: None
    USE_PYTHON3: None

  All features can be prefixed with "Server" or "Client" to limit their
  scope.

  ## Extensions

  Integration tests can be ran for extensions to DIRAC by specifying the
  module name and path such as:

    ./integration_tests.py create --extra-module MyDIRAC=/path/to/MyDIRAC

  This will modify the setup process based on the contents of
  `MyDIRAC/tests/.dirac-ci-config.yaml`. See the Vanilla DIRAC file for the
  available options.

  ## Command completion

  Command competion of typer based scripts can be enabled by running:

    typer --install-completion

  After restarting your terminal you command completion is available using:

    typer ./integration_tests.py run ...

Options:
  --install-completion  Install completion for the current shell.
  --show-completion     Show completion for the current shell, to copy it or
                        customize the installation.

  --help                Show this message and exit.

Commands:
  create               Start a local instance of the integration tests
  destroy              Destroy a local instance of the integration tests
  prepare-environment  Prepare the local environment for installing DIRAC.
  install-server       Install DIRAC in the server container.
  install-client       Install DIRAC in the client container.
  test-server          Run the server integration tests.
  test-client          Run the client integration tests.
  exec-server          Start an interactive session in the server container.
  exec-client          Start an interactive session in the client container.
  list-services        List the services which have been running.
  runsvctrl            Execute runsvctrl inside the server container.
  logs                 Show DIRAC's logs from the service container.
```

BEGINRELEASENOTES

*Core
NEW: DIRAC.Core.Utilities.Extensions module for interacting with extensions in a Python 3 compatible way

*Tests
NEW: ./integration_tests.py helper script for running the integration tests locally

ENDRELEASENOTES

### CODE REVIEW

File: src/DIRAC/Core/Utilities/Extensions.py [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+import sys
+
+import importlib_resources
+import six
+import DIRAC
+
+try:
+  from importlib.machinery import PathFinder
+except ImportError:
+  # Fallback for Python 2
```
Comment (fstagni) [inline src/DIRAC/Core/Utilities/Extensions.py:23]: Just as reminder: we have several cases of using `imp` in DIRAC code, that one day we should address.

File: tests/Jenkins/utilities.sh [hunk 2]
```diff
@@ -318,13 +313,9 @@ installDIRAC() {
...
   fi
 
-  if [[ "${CLIENT_USE_PYTHON3:-}" == "Yes" ]]; then
-    if [[ -n "${DIRACOSVER+x}" ]]; then
-      if [[ "${DIRACOSVER:-}" == "latest" ]]; then
-        DIRACOS2_URL="https://github.com/DIRACGrid/DIRACOS2/releases/latest/download/DIRACOS-Linux-x86_64.sh"
-      else
-        DIRACOS2_URL="https://github.com/DIRACGrid/DIRACOS2/releases/download/${DIRACOSVER}/DIRACOS-Linux-x86_64.sh"
-      fi
+  if [[ "${USE_PYTHON3:-}" == "Yes" ]]; then
```
Comment (fstagni) [inline tests/Jenkins/utilities.sh:316]: what not `${CLIENT_USE_PYTHON3:-}` ?

File: tests/Jenkins/utilities.sh [hunk 3]
```diff
@@ -333,16 +324,15 @@ installDIRAC() {
...
     if [[ -n "${DIRAC_RELEASE+x}" ]]; then
       if [[ -z "${ALTERNATIVE_MODULES}" ]]; then
         pip install "${DIRAC_RELEASE}"
       fi
     fi
     for module_path in "${ALTERNATIVE_MODULES[@]}"; do
-      # ALTERNATIVE_MODULES can be a list of URLs to pip-installable modules
-      # or something like git+https://github.com/fstagni/DIRAC.git@v7r2-fixes33#egg=DIRAC[pilot]
-      pip install "${module_path}"
+      pip install ${PIP_INSTALL_EXTRA_ARGS:-} "${module_path}"
```
Comment (fstagni) [inline tests/Jenkins/utilities.sh:335]: Sorry, can you explain what is `PIP_INSTALL_EXTRA_ARGS` ?

**Comment by fstagni** [comment]
What was wrong with the shell script?

File: src/DIRAC/Core/Utilities/Extensions.py (hunk 1 above)
Comment (chrisburr) [reply src/DIRAC/Core/Utilities/Extensions.py:23]: I've already prepared the commits for v7r3 as part of fixing Python 3 and guiding the functions included in this module.

File: tests/Jenkins/utilities.sh (hunk 2 above)
Comment (chrisburr) [reply tests/Jenkins/utilities.sh:316]: The variables are now automatically stripped of `CLIENT_`/`SERVER_` so you do things like use different DIRACOS versions in the client and server containers or selectively enable other features.

File: tests/Jenkins/utilities.sh (hunk 3 above)
Comment (chrisburr) [reply tests/Jenkins/utilities.sh:335]: At the moment it's only used to pass `-e` if there is a tty attached (i.e. not in CI) so you can edit the code in the services without re-installing or manually copying back and forth with the container. The default can be overridden by passing `--[no-]editable` and it was very convenient when fixing the Python 3 server installation.

File: tests/Jenkins/utilities.sh (hunk 2 above)
Comment (fstagni) [reply tests/Jenkins/utilities.sh:316]: Where is this stripping happening?

Comment (chrisburr) [reply tests/Jenkins/utilities.sh:316]: In the Python script when it generates the `CONFIG` files which are sourced:

https://github.com/DIRACGrid/DIRAC/blob/1dc48d11424a031088afcf5380784fbe8fba53dd/integration_tests.py#L194-L205

**Comment by chrisburr** [comment]
I think it's simpler and easier to maintain if it's in Python, which I think is especially important now it's going to be ran in 4 places (DIRAC/LHCbDIRAC/DIRACOS/DIRACOS2). To do a dump

* I've always found it's integration with the CI to be fragile and on several occasions we've lost a line which causes the tests to not do what we thought they we're. (in fact, currently the threadpool tests do nothing as the variable isn't passed through correctly)
* Testing extensions is painful and requires too much set up. Extensions can now add additional services (e.g. LHCbDIRAC starts an Oracle DB automatically and this could also be useful for testing the webapp, the GitLab CI configuration becomes [effectively one line](https://gitlab.cern.ch/cburr/LHCbDIRAC/-/blob/b5e81f3237232ba4f13114594e65662678a73913/.gitlab-ci.yml#L221) with a [fairly simple config file](https://gitlab.cern.ch/cburr/LHCbDIRAC/-/blob/b5e81f3237232ba4f13114594e65662678a73913/tests/.dirac-ci-config.yaml))
* I've had issues with the use of docker-in-docker causing my docker setup to break until I reboot
* The validation on the parameters passed in is non-existent

In addition the new script makes it possible to have editable installations (for Python 3) so you can reboot the services or re-run the tests without needing to edit the code inside the container. I also included a few convenience functions for things like streaming the logs from all services (see the `--help` output above). Both of these have been quite useful for fixing the Python 3 server installation.

**Comment by fstagni** [comment]
This seems OK but I will finish reviewing once the other PRs are merged. Same for #5000 as I am having a bit of a hard time.

**Comment by fstagni** [comment]
One more thing: this PR needs `importlib_resources` which is merged in DIRACOS master and will be part of DIRACOS v1r19.
DIRACOS v1r19, which will include `arc6`, will be first tested tomorrow in v7r3 hackathon, and not be used for the v7r2 release that will be created today. If the hackathon does not show issues with DIRACOS v1r19, then we'll merge this PR and create a v7r2 release with DIRACOS v1r19.

@atsareg

**Review by fstagni — APPROVED** [review]

**Pull request merged**